File: frontend/lib/models/reminder_log.dart
/// One entry from the reminder_logs table — shows when a WhatsApp nudge was
/// dispatched to the customer or super admin for an installment.
class ReminderLog {
  const ReminderLog({
    required this.id,
    required this.saleId,
    required this.installmentId,
    required this.recipientType,
    required this.recipientPhone,
    required this.channel,
    required this.message,
    required this.dueDate,
    this.sentAt,
    required this.status,
  });

  final int id;
  final int saleId;
  final int? installmentId;
  final String recipientType; // 'customer' | 'super_admin'
  final String recipientPhone;
  final String channel; // 'whatsapp'
  final String message;
  final DateTime dueDate;
  final DateTime? sentAt;
  final String status; // 'pending' | 'sent' | 'failed'

  bool get isSent => status == 'sent';

  factory ReminderLog.fromJson(Map<String, dynamic> j) => ReminderLog(
        id: j['id'] as int,
        saleId: j['sale_id'] as int,
        installmentId: j['installment_id'] as int?,
        recipientType: (j['recipient_type'] as String?) ?? '',
        recipientPhone: (j['recipient_phone'] as String?) ?? '',
        channel: (j['channel'] as String?) ?? 'whatsapp',
        message: (j['message'] as String?) ?? '',
        dueDate:
            DateTime.tryParse((j['due_date'] as String?) ?? '') ?? DateTime.now(),
        sentAt: j['sent_at'] == null
            ? null
            : DateTime.tryParse(j['sent_at'] as String),
        status: (j['status'] as String?) ?? 'pending',
      );
}

File: backend/app/models/reminder_log.py
"""Reminder log — one row per recipient per reminder; sent_at = when dispatched."""
from datetime import date, datetime

from sqlalchemy import BigInteger, Date, ForeignKey, Identity, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db import Base
from app.models.enums import (
    ReminderChannel,
    ReminderRecipient,
    ReminderStatus,
    pg_enum,
)


class ReminderLog(Base):
    __tablename__ = "reminder_logs"

    id: Mapped[int] = mapped_column(BigInteger, Identity(always=True), primary_key=True)
    # nullable so standalone / test sends (not tied to a sale) can be logged too
    sale_id: Mapped[int | None] = mapped_column(
        BigInteger, ForeignKey("sales.id", ondelete="CASCADE")
    )
    installment_id: Mapped[int | None] = mapped_column(
        BigInteger, ForeignKey("sale_installments.id", ondelete="SET NULL")
    )
    recipient_type: Mapped[ReminderRecipient] = mapped_column(
        pg_enum(ReminderRecipient, "reminder_recipient"), nullable=False
    )
    recipient_phone: Mapped[str] = mapped_column(Text, nullable=False)
    channel: Mapped[ReminderChannel] = mapped_column(
        pg_enum(ReminderChannel, "reminder_channel"),
        nullable=False,
        server_default=ReminderChannel.whatsapp.value,
    )
    message: Mapped[str] = mapped_column(Text, nullable=False, server_default="")
    due_date: Mapped[date | None] = mapped_column(Date)
    sent_at: Mapped[datetime | None] = mapped_column()
    status: Mapped[ReminderStatus] = mapped_column(
        pg_enum(ReminderStatus, "reminder_status"),
        nullable=False,
        server_default=ReminderStatus.pending.value,
    )
    provider_message_id: Mapped[str | None] = mapped_column(Text)
    error: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)
